### PR TITLE
Edit "application_scorecard_full_suite_Excel_Final"

### PR DESCRIPTION
**Pull request by LoiAnsah**
# Pull Request Description

## What and why?
Create a notebook for an Excel use-case.

## Checklist
- [ ] What and why
- [ ] Screenshots or videos (Frontend)
- [ ] How to test
- [ ] What needs special review
- [ ] Dependencies, breaking changes, and deployment notes
- [ ] Labels applied
- [ ] PR linked to Shortcut
- [ ] Unit tests added (Backend)
- [ ] Tested locally
- [ ] Documentation updated (if required)
- [ ] Environment variable additions/changes documented (if required)

### CODE REVIEW

**Comment by validbeck** [comment]
@LoiAnsah You should actually get Michael to review this first, not me — since he knows the context and what the outcome should be. :)

**Comment by MichaelIngvarRoenning** [comment]
the rest looks good, I would just add color on defining the model object and assigning predictions (i.e. through the Excel sheet). Remember, Excel is the key here and we want to teach users to translate Excel to ValidMind.

**Review by MichaelIngvarRoenning — APPROVED** [review]

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 1]
```diff
@@ -0,0 +1,1363 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc4_'></a>\n",
+    "\n",
+    "## Initialize the Python environment\n",
+    "\n",
+    "Since we are working with excel files, we need to install the `openpyxl` library:"
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:169]: You need to be more explicit as to why we're performing tasks. You've alluded to it, but you didn't state outright what the usage would be. 

Many users will already be familiar with this library, but part of making writing democratic and accessible is not assuming base knowledge when it would be very simple to outline it instead of being vague.

Here we also want to confirm the outcome of the previous step as the setup there is crucial.

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 2]
```diff
@@ -0,0 +1,1363 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc4_'></a>\n",
+    "\n",
+    "## Initialize the Python environment\n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:167]: This section is also more than just initializing the Python environment — see if you can figure out a better heading here.

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 3]
```diff
@@ -0,0 +1,1363 @@
...
+    "%matplotlib inline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc4_1_'></a>\n",
+    "\n",
+    "### Preview the documentation template\n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:211]: I feel like you copy pasted sections from other notebooks — and did not check if the heading levels made sense after you copied them. Especially since you named the top-level heading just Python set up, does this make sense?

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 4]
```diff
@@ -0,0 +1,1363 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc5_'></a>\n",
+    "\n",
+    "## Load the sample dataset\n",
+    "\n",
+    "You'll need to [download](/CreditRiskData.xlsx) the dataset and load it into a pandas [DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html).\n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:235]: Here, it's better if you make the link to the spreadsheet explicit. Here's an example from a validation notebook I wrote:

Screenshot 2025-07-30 at 4 07 22 PM

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 5]
```diff
@@ -0,0 +1,1363 @@
...
+    "\n",
+    "The dataset is provided as an Excel workbook with five sheets, each representing a different stage of data preparation. These sheets are indexed from 0 to 4:\n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:237]: You could likely combine this sentence and the above with the hint/example I gave to make it more clear.

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 6]
```diff
@@ -0,0 +1,1363 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc5_1_'></a>\n",
+    "\n",
+    "### Prepocess the dataset\n",
+    "\n",
+    "In the preprocessing step we perform a number of operations to get ready for building our application scorecard. A few of those steps include: \n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:434]: This language is both very wordy and doesn't actually tell you why you want to do things! No need for this lead-in fluff.

Do you think there's a guide as to how you can word this better? ;) I feel like I gave you a list of notebooks I wrote that include one on application scorecard models... are you using reference material/researching more when you write?

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 7]
```diff
@@ -0,0 +1,1363 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc5_2_'></a>\n",
+    "\n",
+    "### Feature engineering\n",
+    "\n",
+    "In the feature engineering phase, we apply specific transformations to optimize the dataset for predictive modeling in our application scorecard. \n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:610]: Same comment as the preprocess section. Please give this some more attention.

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 8]
```diff
@@ -0,0 +1,1363 @@
...
+    "fe_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc6_'></a>\n",
+    "\n",
+    "## Bring In Model and Assign Prediction\n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:784]: Why does the casing of this heading not match our style guide or any of the other headings? :(

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 9]
```diff
@@ -0,0 +1,1363 @@
...
+    "\n",
+    "In this step, we prepare the model using feature-engineered data already split into training (**Train Data**, index `3`) and testing (**Test Data**, index `4`) sets. We'll load those datasets into pandas DataFrames:\n"
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:786]: Again, still very fluffy — please edit these down to have less filler.

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 10]
```diff
@@ -0,0 +1,1363 @@
...
+    "test_df = pd.read_excel('CreditRiskData.xlsx', sheet_name=4,engine='openpyxl')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a id='toc7_'></a>\n",
+    "\n",
+    "## Document the model\n",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:806]: This information feels like it's better introduced at the top of the notebook, no? Please take a look at the reference material provided for some guidance as to how to clean up this big section — The "ValidMind for model development" series is a great tool here.

File: notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb [hunk 11]
```diff
@@ -0,0 +1,1363 @@
...
+     "text": [
+      "validmind.model_validation.sklearn.OverfitDiagnosis\n",
+      "validmind.model_validation.sklearn.RobustnessDiagnosis\n",
+      "Error running test validmind.model_validation.sklearn.RobustnessDiagnosis: dummy() takes 0 positional arguments but 1 was given\n",
+      "validmind.model_validation.FeaturesAUC\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
```
Comment (validbeck) [inline notebooks/code_samples/credit_risk/application_scorecard_full_suite_Excel_Final.ipynb:1237]: Did you mean to include the output here? It doesn't seem like it. The output looks wrong, anywhow.

**Comment by LoiAnsah** [comment]
Hi @validbeck,
I understand I won’t be able to make further changes, but I hope the ones I’ve submitted are close to your expectations. Thanks again for your feedback and support throughout!

**Review by validbeck — APPROVED** [review]
I cleaned up the notebook a bit more after @LoiAnsah:

- Shuffled around the headings to be cleaner
- Finetuned the section where we load the sample datasets — in particular **we don't actually call the FE dataset anywhere as we've already got the split test & train datasets in the excel sheet so I removed that from the spreadsheet and any references in the notebook**
- For running tests, **there was a redundancy in how we set the test configurations so I just streamlined this section — I also grouped the kinds of tests we run with which datasets** and explained a bit more why we structure things the way we do in the calls
- For the what's next section, I added some context for the individual tests that don't get automatically logged to the template

@MichaelIngvarRoenning if you want to take a look at the changes before I smush the notebook into `main`:

| Ama's Version | Updated Version |
|--|--|
| [Ama's Last Version](https://github.com/validmind/validmind-library/blob/e557ad01191e19e4823999ee9eab835c453f4950/notebooks/code_samples/credit_risk/document_excel_application_scorecard.ipynb) | [LATEST VERSION LINK](https://github.com/validmind/validmind-library/blob/ama8636/sc-11408/edit-document-an-application-scorecard-model/notebooks/code_samples/credit_risk/document_excel_application_scorecard.ipynb) |

**Comment by MichaelIngvarRoenning** [comment]
everything look good but is it possible to somehow bring the below code as a function we bring in or something in that effect so we can hide the "complicated" looking code.

def run_doc_tests(test_config):
    for t in test_config:
        print(t)
        try:
            test = test_config[t]

            # Check if test uses input_grid
            if 'input_grid' in test:
                if 'params' in test:
                    vm.tests.run_test(
                        t,
                        input_grid=test['input_grid'],
                        params=test['params'],
                        show=False
                    ).log()
                else:
                    vm.tests.run_test(
                        t,
                        input_grid=test['input_grid'],
                        show=False
                    ).log()

            # Fallback to regular input-based test
            else:
                if 'params' in test:
                    vm.tests.run_test(
                        t,
                        inputs=test['inputs'],
                        params=test['params'],
                        show=False
                    ).log()
                else:
                    vm.tests.run_test(
                        t,
                        inputs=test['inputs'],
                        show=False
                    ).log()

        except Exception as e:
            print(f"Error running test {t}: {str(e)}")

**Comment by validbeck** [comment]
@MichaelIngvarRoenning 

> everything look good but is it possible to somehow bring the below code as a function we bring in or something in that effect so we can hide the "complicated" looking code.

This is your original script, I just prettified it and removed the `test_config2` redundancy (see below) — AFAK we need to be able to iterate over the tests with this function. 🤔 

Screenshot 2025-08-08 at 11 23 24 AM

**Comment by validbeck** [comment]
@MichaelIngvarRoenning I ended up cleaning up the function so it looks less long at least:

Screenshot 2025-08-08 at 12 04 44 PM

**Comment by validbeck** [comment]
I also removed the `validmind.model_validation.sklearn.RobustnessDiagnosis` test as it was throwing an error: 

```bash
Error running test validmind.model_validation.sklearn.RobustnessDiagnosis: dummy() takes 0 positional arguments but 1 was given
validmind.model_validation.FeaturesAUC
```

I figured we had enough examples.

**Comment by MichaelIngvarRoenning** [comment]
> @MichaelIngvarRoenning I ended up cleaning up the function so it looks less long at least:
> 
> Screenshot 2025-08-08 at 12 04 44 PM

Looks great!
